nexus/src/app/crucible.rs

```diff
@@ -0,0 +1,1051 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Functions common to interacting with Crucible agents
+
+use super::*;
+
+use anyhow::anyhow;
+use crucible_agent_client::types::CreateRegion;
+use crucible_agent_client::types::GetSnapshotResponse;
+use crucible_agent_client::types::Region;
+use crucible_agent_client::types::RegionId;
+use crucible_agent_client::types::State as RegionState;
+use crucible_agent_client::Client as CrucibleAgentClient;
+use futures::StreamExt;
+use nexus_db_queries::db;
+use nexus_db_queries::db::identity::Asset;
+use omicron_common::api::external::Error;
+use omicron_common::backoff::{self, BackoffError};
+use omicron_common::progenitor_operation_retry::ProgenitorOperationRetry;
+use omicron_common::progenitor_operation_retry::ProgenitorOperationRetryError;
+use slog::Logger;
+
+// Arbitrary limit on concurrency, for operations issued on multiple regions
+// within a disk at the same time.
+const MAX_CONCURRENT_REGION_REQUESTS: usize = 3;
+
+/// Provides a way for (with BackoffError) Permanent errors to have a different error type than
+/// Transient errors.
+#[derive(Debug, thiserror::Error)]
+enum WaitError {
+    #[error("Transient error: {0}")]
+    Transient(#[from] anyhow::Error),
+
+    #[error("Permanent error: {0}")]
+    Permanent(#[from] Error),
+}
+
+impl super::Nexus {
```

Comment:
I think we really want to get away from having a shared nexus type with all functions hanging off it. This has already made it quite difficult to pass pieces into background tasks by themselves. I'm not sure we should worry about it in this PR, but eventually we are going to end up ripping all this out and turning it into it's own distinct type I believe. Maybe we should get ahead of that now and provide a `CrucibleAgentClient` type that takes specific parameters for what it needs. That may be easier said than done given the need to use the `datastore`, but I figured I'd bring it up for completness and to also get us thinking about how we move on past the chads hanging off Nexus.

CC @david-crespo @jgallagher @hawkw @davepacheco

Reply:
Yeah, I would really like to move away from this as much as we're able to...

Reply:
Without a motivating case of sharing functionality with a background task I'm not sure I want to separate parts of this out yet. I'm definitely glad to do the work in the future that you're talking about.

Reply:
That works for me.